substrate/frame/sassafras/src/lib.rs

```diff
@@ -397,182 +407,84 @@ pub mod pallet {
+				let ticket_id_input = vrf::ticket_id_input(&randomness, envelope.attempt);
 
 				// Check threshold constraint
 				let ticket_id = vrf::make_ticket_id(&ticket_id_input, &ticket_id_pre_output);
-				if ticket_id >= ticket_threshold {
-					debug!(target: LOG_TARGET, "Ignoring ticket over threshold ({:032x} >= {:032x})", ticket_id, ticket_threshold);
-					continue
-				}
+				trace!(target: LOG_TARGET, "Checking ticket {:?}", ticket_id);
 
-				// Check for duplicates
-				if TicketsData::<T>::contains_key(ticket_id) {
-					debug!(target: LOG_TARGET, "Ignoring duplicate ticket ({:032x})", ticket_id);
-					continue
+				if ticket_id >= ticket_threshold {
+					debug!(target: LOG_TARGET, "Ticket over threshold ({:?} >= {:?})", ticket_id, ticket_threshold);
+					return Err(Error::<T>::TicketOverThreshold.into())
 				}
 
 				// Check ring signature
-				let sign_data = vrf::ticket_body_sign_data(&ticket.body, ticket_id_input);
-				if !ticket.signature.ring_vrf_verify(&sign_data, &verifier) {
-					debug!(target: LOG_TARGET, "Proof verification failure for ticket ({:032x})", ticket_id);
-					continue
+				let sign_data = vrf::ticket_id_sign_data(ticket_id_input, &envelope.extra);
+				if !envelope.signature.ring_vrf_verify(&sign_data, &verifier) {
+					debug!(target: LOG_TARGET, "Proof verification failure for ticket ({:?})", ticket_id);
+					return Err(Error::<T>::TicketBadProof.into())
 				}
 
-				if let Ok(_) = valid_tickets.try_push(ticket_id).defensive_proof(
-					"Input segment has same length as bounded destination vector; qed",
-				) {
-					TicketsData::<T>::set(ticket_id, Some(ticket.body));
-				}
+				candidates.push(TicketBody {
+					id: ticket_id,
+					attempt: envelope.attempt,
+					extra: envelope.extra,
+				});
 			}
 
-			if !valid_tickets.is_empty() {
-				Self::append_tickets(valid_tickets);
-			}
+			Self::deposit_tickets(candidates)?;
 
-			Ok(Pays::No.into())
-		}
-
-		/// Plan an epoch configuration change.
-		///
-		/// The epoch configuration change is recorded and will be announced at the beginning
-		/// of the next epoch together with next epoch authorities information.
-		/// In other words, the configuration will be enacted one epoch later.
-		///
-		/// Multiple calls to this method will replace any existing planned config change
-		/// that has not been enacted yet.
-		#[pallet::call_index(1)]
-		#[pallet::weight(T::WeightInfo::plan_config_change())]
-		pub fn plan_config_change(
-			origin: OriginFor<T>,
-			config: EpochConfiguration,
-		) -> DispatchResult {
-			ensure_root(origin)?;
-
-			ensure!(
-				config.redundancy_factor != 0 && config.attempts_number != 0,
-				Error::<T>::InvalidConfiguration
-			);
-			PendingEpochConfigChange::<T>::put(config);
 			Ok(())
 		}
 	}
 
-	#[pallet::validate_unsigned]
-	impl<T: Config> ValidateUnsigned for Pallet<T> {
+	#[pallet::inherent]
+	impl<T: Config> ProvideInherent for Pallet<T> {
```

Comment:
I left it to the default impl:
```rust
	/// - `Ok(None)` indicates that this inherent is not required in this block. The default
	/// implementation returns this.
```
Inherents with tickets are never required. If a node has some tickets to share then he do it, but you don't know and is fine to don't know.

Maybe we can discuss about the `DispatchClass`: 
```rust
		#[pallet::call_index(0)]
		#[pallet::weight((
			T::WeightInfo::submit_tickets(envelopes.len() as u32),
			DispatchClass::Mandatory
		))]
		pub fn submit_tickets(origin: OriginFor<T>, envelopes: TicketsVec<T>) -> DispatchResult {
```

Currently I've set it as `Mandatory`.

https://github.com/paritytech/polkadot-sdk/blob/a164639f7f1223634fb01cf38dab49622ab940ef/substrate/frame/support/src/dispatch.rs#L172

Even though is said to avoid this for "heavy" operations. Even though this tickets verification is not the lightest of the operations, should be said that this kind of inherent is considered central to the consensus algorithm and if is there must be processed.
Should be said that there is an upper bound to the Tickets that can be processed in a single call, so the exact cost can be predicted (verification cost per ticket ~4ms on my machine https://github.com/davxy/crypto-benches/blob/main/vrf/README.md#verify).